# Data Preprocessing

In [1]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
# Importing the dataset
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

In [3]:
# Encoding categorical data
# Encoding the Independent Variable
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]

C:\Users\SHADAAB\AppData\Local\conda\conda\envs\hello-tf\lib\site-packages\sklearn\preprocessing\_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
C:\Users\SHADAAB\AppData\Local\conda\conda\envs\hello-tf\lib\site-packages\sklearn\preprocessing\_encoders.py:392: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [4]:
#But now we will remove one dummy variable here to avoid falling into the dummy variable trap.

In [5]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)


In [6]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# ANN

In [7]:
# Importing the Keras libraries and packages
import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


In [8]:
# Initialising the ANN
classifier = Sequential()

In [9]:
# Adding the input layer and the first hidden layer
classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))
#relu=rectifier transfer function     
#units=no. of hidden layer. input+output/2.  11+1/2=6

Instructions for updating:
Colocations handled automatically by placer.


In [10]:
# Adding the second hidden layer
classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))

In [11]:
# Adding the output layer
classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
#unit= 3 if you have 3 categories for dependent variable
#activation= 'softmax' softmax is actually the sigmoid function, but applied to a dependent variable that has more than two
#categories

# But here, we have two categories, two classes, so we are fine with the sigmoid function

In [12]:
# We are done adding the layers of our artificial neural network, and now  we're gonna do is to compile the whole 
# artificial neural network, that is basically applying stochastic gradient descent on the whole artificial neural network.
# That's what compiling means.

In [13]:
# Compiling the ANN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [14]:
# Fitting the ANN to the Training set
classifier.fit(X_train, y_train, batch_size = 10, epochs = 100)

Instructions for updating:
Use tf.cast instead.
Epoch 1/100
8000/8000 [==============================] - 3s 330us/step - loss: 0.4922 - acc: 0.7952
Epoch 2/100
8000/8000 [==============================] - 1s 167us/step - loss: 0.4284 - acc: 0.7960
Epoch 3/100
8000/8000 [==============================] - 1s 166us/step - loss: 0.4236 - acc: 0.7960
Epoch 4/100
8000/8000 [==============================] - 1s 168us/step - loss: 0.4191 - acc: 0.8156
Epoch 5/100
8000/8000 [==============================] - 1s 174us/step - loss: 0.4171 - acc: 0.8262
Epoch 6/100
8000/8000 [==============================] - 2s 193us/step - loss: 0.4150 - acc: 0.8285
Epoch 7/100
8000/8000 [==============================] - 2s 219us/step - loss: 0.4130 - acc: 0.8309 1s
Epoch 8/100
8000/8000 [==============================] - 2s 188us/step - loss: 0.4099 - acc: 0.8332
Epoch 9/100
8000/8000 [==============================] - 1s 185us/step - loss: 0.4078 - acc: 0.8335
Epoch 10/100
8000/8000 [=========================

# Making predictions and evaluating the model


In [15]:
# Predicting the Test set results
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

In [16]:
# Making the Confusion Matrix and classification report
from sklearn.metrics import confusion_matrix, classification_report
print(confusion_matrix(y_test, y_pred))
print('\n')
print(classification_report(y_test,y_pred))

[[1515   80]
 [ 207  198]]


              precision    recall  f1-score   support

           0       0.88      0.95      0.91      1595
           1       0.71      0.49      0.58       405

   micro avg       0.86      0.86      0.86      2000
   macro avg       0.80      0.72      0.75      2000
weighted avg       0.85      0.86      0.85      2000



In [ ]:
# The accuracy is the number of correct predictions divided by the total number of predictions
# here correct predictions- 1539,146

In [17]:
(1515+198)/2000   #accuracy

0.8565

# Accuracy of our model is 85.6%

So, eventually, we can validate our model, and now the bank can use it to make a ranking of their customers, 
ranked by their probability to leave the bank, from the customer that has the highest probability to leave the bank,
down to the customer that has the lowest probability to leave the bank.

# Predicting a single new observation


In [18]:
"""Predict if the customer with the following informations will leave the bank:
Geography: France
Credit Score: 600
Gender: Male
Age: 40
Tenure: 3
Balance: 60000
Number of Products: 2
Has Credit Card: Yes
Is Active Member: Yes
Estimated Salary: 50000"""

'Predict if the customer with the following informations will leave the bank:\nGeography: France\nCredit Score: 600\nGender: Male\nAge: 40\nTenure: 3\nBalance: 60000\nNumber of Products: 2\nHas Credit Card: Yes\nIs Active Member: Yes\nEstimated Salary: 50000'

In [19]:
new_prediction = classifier.predict(sc.transform(np.array([[0.0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])))
new_prediction = (new_prediction > 0.5)

In [20]:
#Making the confusion matrix
print(confusion_matrix(y_test,y_pred))
print('\n')
print(classification_report(y_test,y_pred))

[[1515   80]
 [ 207  198]]


              precision    recall  f1-score   support

           0       0.88      0.95      0.91      1595
           1       0.71      0.49      0.58       405

   micro avg       0.86      0.86      0.86      2000
   macro avg       0.80      0.72      0.75      2000
weighted avg       0.85      0.86      0.85      2000



In [21]:
new_prediction

array([[False]])

# Therefore our customer doesn't leave the bank

------------------

# Evaluating, Improving and Tuning ANN

In [22]:
#Evaluating the ANN
from keras.wrappers.scikit_learn import KerasClassifier

In [23]:
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense

In [24]:
def build_classifier():
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
    classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier

In [25]:
classifier = KerasClassifier(build_fn = build_classifier, batch_size = 10, epochs = 100)
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10, n_jobs = -1) 
                                                #n_jobs=-1 (multiple CPU)use all the CPU's and therefore what it will do 
                                               # is run parallel computations to get this done faster.


In [27]:
mean = accuracies.mean()
variance = accuracies.std()

In [28]:
print('mean=',mean)
print('\n')
print('Variance=',variance)

mean= 0.8361249949783087


Variance= 0.010627939883213927


Overfitting is when our model was trained too much on the training set, too much that it becomes much less performance on the 
test set and we can observe this when we have a large difference of accuracies between training set and the test set.

Generally, when overfitting happens, we have a much higher accuracy on the training set than the test set. 

And another way to detect overfitting is when we observe a high variance when applying k-fold cross-validation because indeed, 
when it's overfitted on the training set, that is when our model learn too much, well sometimes it will succeed on new 
observations in one test set, when the correlations are similar to what the model learned too much.

And sometimes our model won't succeed on other test sets because the correlations that it learned too much, unfortunately 
don't apply to these test sets. And therefore, in the vector of accuracies that is obtained with k-fold cross-validations, 
we have some high accuracies and some low accuracies and therefore we have high variance and that's how we detect overfitting.

# Our variance is 0.0106, which is less. our accuracy is with K-fold validation. If in case there is some difference in variance then we need to go for drop regularization.
# But here, our variance is less so no need to go for drop regularization

Where do we need to apply dropout to our artificial neural network and how does it work?

At each iteration of the training some neurons of our artificial neural network are randomly disabled to prevent them from being too dependent on each other when they learn the correlations and therefore, by over-writing these neurons the artificial neural network learns several independent correlations in the data, because each time there is not the same configuration
of the neurons.
And the fact that we get these independent correlations of the data, thanks to the fact that the neurons work more independently, that prevents the neurons from learning too much and therefore that prevents overfitting.

In [29]:
"""Dropout Regularization to reduce overfitting if needed
from kera.layers import Dropout

# Initialising the ANN
classifier = Sequential()

# Adding the input layer and the first hidden layer
classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))
# classifier.add(Dropout(p = 0.1))

# Adding the second hidden layer
classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
# classifier.add(Dropout(p = 0.1))

# Adding the output layer
classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))

"""

"Dropout Regularization to reduce overfitting if needed\nfrom kera.layers import Dropout\n\n# Initialising the ANN\nclassifier = Sequential()\n\n# Adding the input layer and the first hidden layer\nclassifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))\n# classifier.add(Dropout(p = 0.1))\n\n# Adding the second hidden layer\nclassifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))\n# classifier.add(Dropout(p = 0.1))\n\n# Adding the output layer\nclassifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))\n\n"

In [30]:
# Tuning the ANN
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Dense

In [31]:
def build_classifier(optimizer):
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
    classifier.compile(optimizer = optimizer, loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier

In [34]:
classifier = KerasClassifier(build_fn = build_classifier)
parameters = {'batch_size': [25, 32],
              'epochs': [100, 500],
              'optimizer': ['adam', 'rmsprop']}
grid_search = GridSearchCV(estimator = classifier,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,n_jobs=-1)

In [35]:
grid_search = grid_search.fit(X_train, y_train)
best_parameters = grid_search.best_params_
best_accuracy = grid_search.best_score_

Epoch 1/500
8000/8000 [==============================] - 13s 2ms/step - loss: 0.5432 - acc: 0.7962
Epoch 2/500
8000/8000 [==============================] - 1s 111us/step - loss: 0.4337 - acc: 0.7960
Epoch 3/500
8000/8000 [==============================] - 1s 108us/step - loss: 0.4283 - acc: 0.7960
Epoch 4/500
8000/8000 [==============================] - 1s 103us/step - loss: 0.4246 - acc: 0.7960
Epoch 5/500
8000/8000 [==============================] - 1s 100us/step - loss: 0.4214 - acc: 0.7960
Epoch 6/500
8000/8000 [==============================] - 1s 115us/step - loss: 0.4188 - acc: 0.8112
Epoch 7/500
8000/8000 [==============================] - 1s 103us/step - loss: 0.4169 - acc: 0.8256
Epoch 8/500
8000/8000 [==============================] - 1s 92us/step - loss: 0.4158 - acc: 0.8275
Epoch 9/500
8000/8000 [==============================] - 1s 98us/step - loss: 0.4144 - acc: 0.8294
Epoch 10/500
8000/8000 [==============================] - 1s 98us/step - loss: 0.4131 - acc: 0.8314
Epo

8000/8000 [==============================] - 1s 90us/step - loss: 0.3970 - acc: 0.8376
Epoch 164/500
8000/8000 [==============================] - 1s 88us/step - loss: 0.3967 - acc: 0.8377
Epoch 165/500
8000/8000 [==============================] - 1s 87us/step - loss: 0.3967 - acc: 0.8366
Epoch 166/500
8000/8000 [==============================] - 1s 92us/step - loss: 0.3969 - acc: 0.8374
Epoch 167/500
8000/8000 [==============================] - 1s 89us/step - loss: 0.3971 - acc: 0.8371
Epoch 168/500
8000/8000 [==============================] - 1s 94us/step - loss: 0.3971 - acc: 0.8369
Epoch 169/500
8000/8000 [==============================] - 1s 93us/step - loss: 0.3969 - acc: 0.8382
Epoch 170/500
8000/8000 [==============================] - 1s 91us/step - loss: 0.3973 - acc: 0.8375
Epoch 171/500
8000/8000 [==============================] - 1s 93us/step - loss: 0.3967 - acc: 0.8374
Epoch 172/500
8000/8000 [==============================] - 1s 90us/step - loss: 0.3969 - acc: 0.8376
Epoc

8000/8000 [==============================] - 1s 93us/step - loss: 0.3964 - acc: 0.8365
Epoch 245/500
8000/8000 [==============================] - 1s 91us/step - loss: 0.3965 - acc: 0.8366
Epoch 246/500
8000/8000 [==============================] - 1s 93us/step - loss: 0.3964 - acc: 0.8387
Epoch 247/500
8000/8000 [==============================] - 1s 95us/step - loss: 0.3966 - acc: 0.8372
Epoch 248/500
8000/8000 [==============================] - 1s 93us/step - loss: 0.3963 - acc: 0.8365
Epoch 249/500
8000/8000 [==============================] - 1s 90us/step - loss: 0.3961 - acc: 0.8379
Epoch 250/500
8000/8000 [==============================] - 1s 88us/step - loss: 0.3965 - acc: 0.8372
Epoch 251/500
8000/8000 [==============================] - 1s 87us/step - loss: 0.3967 - acc: 0.8354
Epoch 252/500
8000/8000 [==============================] - 1s 89us/step - loss: 0.3966 - acc: 0.8376
Epoch 253/500
8000/8000 [==============================] - 1s 86us/step - loss: 0.3966 - acc: 0.8367
Epoc

Epoch 405/500
8000/8000 [==============================] - 1s 88us/step - loss: 0.3956 - acc: 0.8367
Epoch 406/500
8000/8000 [==============================] - 1s 86us/step - loss: 0.3956 - acc: 0.8347
Epoch 407/500
8000/8000 [==============================] - 1s 89us/step - loss: 0.3958 - acc: 0.8377
Epoch 408/500
8000/8000 [==============================] - 1s 86us/step - loss: 0.3954 - acc: 0.8364
Epoch 409/500
8000/8000 [==============================] - 1s 85us/step - loss: 0.3961 - acc: 0.8364
Epoch 410/500
8000/8000 [==============================] - 1s 89us/step - loss: 0.3959 - acc: 0.8377
Epoch 411/500
8000/8000 [==============================] - 1s 86us/step - loss: 0.3956 - acc: 0.8364
Epoch 412/500
8000/8000 [==============================] - ETA: 0s - loss: 0.3948 - acc: 0.837 - 1s 103us/step - loss: 0.3954 - acc: 0.8364
Epoch 413/500
8000/8000 [==============================] - 1s 88us/step - loss: 0.3957 - acc: 0.8370
Epoch 414/500
8000/8000 [===========================

In [36]:
print('best_parameters=',best_parameters)
print('best_accuracy=',best_accuracy)

best_parameters= {'batch_size': 25, 'epochs': 500, 'optimizer': 'adam'}
best_accuracy= 0.8485


## And now after evaluating, improving and tuning the ANN, our model's Accuracy is 84.8%